bundles/org.openhab.binding.upb/src/main/java/org/openhab/binding/upb/handler/SerialIoThread.java

```diff
@@ -0,0 +1,284 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.upb.handler;
+
+import static java.nio.charset.StandardCharsets.US_ASCII;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Arrays;
+import java.util.TooManyListenersException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.smarthome.io.transport.serial.SerialPort;
+import org.eclipse.smarthome.io.transport.serial.SerialPortEvent;
+import org.eclipse.smarthome.io.transport.serial.SerialPortEventListener;
+import org.openhab.binding.upb.handler.UPBIoHandler.CmdStatus;
+import org.openhab.binding.upb.internal.message.MessageBuilder;
+import org.openhab.binding.upb.internal.message.UPBMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Event loop for serial communications. Handles sending and receiving UPB messages.
+ *
+ * @author Marcus Better - Initial contribution
+ *
+ */
+public class SerialIoThread extends Thread implements SerialPortEventListener {
+    private static final int WRITE_QUEUE_LENGTH = 128;
+    private static final int ACK_TIMEOUT_MS = 500;
+
+    private final Logger logger = LoggerFactory.getLogger(SerialIoThread.class);
+    private final byte[] ENABLE_MESSAGE_MODE_CMD = { 0x17, 0x70, 0x02, (byte) 0x8e, 0x0d };
+    private final byte[] buffer = new byte[512];
+    private final MessageListener listener;
+    private final ExecutorService writeExecutor = new ThreadPoolExecutor(1, 1, 30, TimeUnit.SECONDS,
+            new LinkedBlockingQueue<>(WRITE_QUEUE_LENGTH));
+
+    private int bufferLength = 0;
+    private volatile WriteRunnable currentWrite = null;
+    private volatile SerialPort serialPort;
+    private volatile boolean done;
+
+    public SerialIoThread(final SerialPort serialPort, final MessageListener listener) {
+        this.serialPort = serialPort;
+        this.listener = listener;
+    }
+
+    @Override
+    public void serialEvent(final SerialPortEvent event) {
+        try {
+            logger.trace("RXTX library CPU load workaround, sleep forever");
+            Thread.sleep(Long.MAX_VALUE);
+        } catch (final InterruptedException e) {
+            // ignore
+        }
+    }
+
+    @Override
+    public void run() {
+        // RXTX serial port library causes high CPU load
+        // Start event listener, which will just sleep and slow down event loop
+        try {
+            serialPort.addEventListener(this);
+        } catch (final TooManyListenersException e) {
+            throw new RuntimeException(e);
```

Comment:
Instead of throwing a RuntimeException you should be logging then error and then return immediately.